src/execution/join_hashtable.cpp

```diff
@@ -275,32 +270,156 @@ void JoinHashTable::Finalize() {
-				key_locations[i] = dataptr;
-				dataptr += entry_size;
+			idx_t entries_count = MinValue<idx_t>(STANDARD_VECTOR_SIZE, block.count - entry);
+			for (idx_t i = 0; i < entries_count; i++) {
+				hash_data[i] = Load<hash_t>((data_ptr_t)(entries_buffer + pointer_offset));
+				key_locations[i] = entries_buffer;
+				entries_buffer += entry_size; // walk to next tuple
 			}
-			// now insert into the hash table
-			InsertHashes(hashes, next, key_locations);
-
-			entry += next;
+			// now insert the key_locations into the hash table
+			InsertHashes(hashes, entries_count, key_locations);
+			entry += entries_count; // walk entries_count
 		}
 		pinned_handles.push_back(move(handle));
 	}
-
+	// In case of primary key with an inner join do a semi-join instead
+	if (join_type == JoinType::INNER && has_primary_key) {
+		join_type = JoinType::SEMI;
+	}
 	finalized = true;
 }
 
+void JoinHashTable::InsertHashes(Vector &hashes, idx_t count_tuples, data_ptr_t key_locations[]) {
+	D_ASSERT(hashes.GetType().id() == LogicalTypeId::HASH);
+
+	// create a vector for indices and apply the bitmask in the hash_values
+	ApplyBitmask(hashes, count_tuples);
+	hashes.Normalify(count_tuples); // Flatten the vector
+	D_ASSERT(hashes.GetVectorType() == VectorType::FLAT_VECTOR);
+	auto pointers = (data_ptr_t *)hash_map->node->buffer;
+	auto indices = FlatVector::GetData<hash_t>(hashes);
+	// First, fill the hash_map and handle conflicts with a next_pointer
+	if (has_primary_key && join_type == JoinType::INNER) {
+		// For inner joins we check whether the build is a primary key during the insertion into the hashmap
+		InsertHashesAndCheckPrimaryKey(count_tuples, indices, key_locations, pointers); // inlined
+	} else {
+		for (idx_t i = 0; i < count_tuples; i++) {
+			// For each tuple, the hash_value will be replaced by a pointer to the next_entry in the hash_map
+			auto index = indices[i];
+			auto next_ptr = key_locations[i] + pointer_offset;
+			// replace the hash_value in the current entry and point to a position in the hash_map
+			Store<data_ptr_t>(pointers[index], next_ptr);
+			// store the pointer to the current tuple entry in the hash_map
+			pointers[index] = key_locations[i];
+		}
+	}
+}
+
+void JoinHashTable::InsertHashesAndCheckPrimaryKey(idx_t count_tuples, hash_t *indices, data_ptr_t key_locations[],
+                                                   data_ptr_t *pointers) {
+	auto col_offsets = layout.GetOffsets();
+	vector<data_ptr_t> conflict_entries;
+	conflict_entries.reserve(count_tuples);
+	for (idx_t i = 0; i < count_tuples; i++) {
+		// For each tuple, the hash_value will be replaced by a pointer to the next_entry in the hash_map
+		auto index = indices[i];
+		auto next_ptr = key_locations[i] + pointer_offset;
+		// In case this is still a primary key and there is a conflict
+		if (has_primary_key && pointers[index] != nullptr) {
+			// for each key pair in the entry
+			for (idx_t key_idx = 0; key_idx != condition_types.size(); ++key_idx) {
```

Comment:
This runs the key comparison in a non-vectorized manner, would it not make sense to gather all the locations of the keys we need to compare and do the comparison in a vectorized manner?